lib/teleterm/services/remoteclientcache/remoteclientcache.go

```diff
@@ -0,0 +1,178 @@
+		//
+		// If the cert is expired, we will remove the client and make an attempt
+		// to connect to proxy which will fail with an appropriate error.
+		proxyClient := c.getFromCache(clusterURI)
+		if proxyClient != nil {
+			if cluster.Connected() {
+				return proxyClient, nil
+			}
+			err := c.InvalidateForRootCluster(clusterURI.GetRootClusterURI())
+			if err != nil {
+				c.log.WithError(err).Errorf("Failed to invalidate expired remote client for %q.", clusterURI)
+			}
+		}
+
+		newProxyClient, err := clusterClient.ConnectToProxy(ctx)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		// We'll save the remote client in the cache, so we don't have to
+		// build a new connection next time.
+		// All remote clients will be closed when the daemon exits.
+		err = c.addToCache(clusterURI, newProxyClient)
+		if err != nil {
+			c.log.WithError(err).Errorf("Failed to add remote client for %q to cache.", clusterURI)
+		} else {
+			c.log.Infof("Added remote client for %q to cache.", clusterURI)
+		}
+
+		return newProxyClient, nil
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	clt, ok := cltI.(*client.ProxyClient)
+	if !ok {
+		return nil, trace.BadParameter("unexpected type %T received for proxy client", cltI)
+	}
+
+	return clt, nil
+}
+
+func (c *Cache) InvalidateForRootCluster(rootClusterURI uri.ResourceURI) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	errors := make([]error, 0, len(c.clients))
+	deleted := make([]uri.ResourceURI, 0, len(c.clients))
+
+	for resourceURI, clt := range c.clients {
+		if resourceURI.GetRootClusterURI() == rootClusterURI {
+			err := clt.Close()
+			if err != nil {
+				errors = append(errors, err)
+			}
+			deleted = append(deleted, resourceURI.GetClusterURI())
+			delete(c.clients, resourceURI)
+		}
+	}
+
+	c.log.Infof("Invalidated cached remote clients for root cluster %q: %v", rootClusterURI, deleted)
+
+	return trace.NewAggregate(errors...)
+
+}
+
+func (c *Cache) Close() error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	errors := make([]error, 0, len(c.clients))
+	for _, clt := range c.clients {
+		errors = append(errors, clt.Close())
+	}
+	c.clients = make(map[uri.ResourceURI]*client.ProxyClient)
+
+	return trace.NewAggregate(errors...)
+}
+
+func (c *Cache) addToCache(clusterURI uri.ResourceURI, proxyClient *client.ProxyClient) error {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	var err error
+	if c.clients[clusterURI] != nil {
+		err = c.clients[clusterURI].Close()
+	}
+	c.clients[clusterURI] = proxyClient
+
+	go func() {
+		err := proxyClient.Client.Wait()
+		if err != nil {
+			c.mu.Lock()
+			defer c.mu.Unlock()
+
+			delete(c.clients, clusterURI)
+			c.log.WithError(err).Infof("Remote client to %q has been closed and removed from cache.", clusterURI)
+		}
+	}()
+	return trace.Wrap(err)
+}
+
+func (c *Cache) getFromCache(clusterURI uri.ResourceURI) *client.ProxyClient {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	remoteClt := c.clients[clusterURI]
```

Comment:
In web sessions' cache we also compare remote site: https://github.com/gravitational/teleport/blob/2baa7e72d9c22fc5263d46a03a4ba1e17de74ffd/lib/web/sessions.go#L1024. Why is it needed and when it can change? Do we need here it too?